cni/network/invoker_cns.go

```diff
@@ -70,94 +72,146 @@ func (invoker *CNSIPAMInvoker) Add(addConfig IPAMAddConfig) (IPAMAddResult, erro
-		ncGatewayIPAddress: response.PodIpInfo.NetworkContainerPrimaryIPConfig.GatewayIPAddress,
-		hostSubnet:         response.PodIpInfo.HostPrimaryIPInfo.Subnet,
-		hostPrimaryIP:      response.PodIpInfo.HostPrimaryIPInfo.PrimaryIP,
-		hostGateway:        response.PodIpInfo.HostPrimaryIPInfo.Gateway,
-	}
+	addResult := IPAMAddResult{}
 
-	// set the NC Primary IP in options
-	addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+	for i := 0; i < len(response.PodIPInfo); i++ {
+		info := IPResultInfo{
+			podIPAddress:       response.PodIPInfo[i].PodIPConfig.IPAddress,
+			ncSubnetPrefix:     response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.PrefixLength,
+			ncPrimaryIP:        response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.IPSubnet.IPAddress,
+			ncGatewayIPAddress: response.PodIPInfo[i].NetworkContainerPrimaryIPConfig.GatewayIPAddress,
+			hostSubnet:         response.PodIPInfo[i].HostPrimaryIPInfo.Subnet,
+			hostPrimaryIP:      response.PodIPInfo[i].HostPrimaryIPInfo.PrimaryIP,
+			hostGateway:        response.PodIPInfo[i].HostPrimaryIPInfo.Gateway,
+		}
 
-	log.Printf("[cni-invoker-cns] Received info %+v for pod %v", info, podInfo)
+		// set the NC Primary IP in options
+		// SNATIPKey set is not for ipv6
+		if net.ParseIP(info.ncPrimaryIP).To4() != nil {
+			addConfig.options[network.SNATIPKey] = info.ncPrimaryIP
+		}
 
-	// set result ipconfigArgument from CNS Response Body
-	ip, ncipnet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
-	if ip == nil {
-		return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
-	}
+		log.Printf("[cni-invoker-cns] Received info %+v for pod %v", info, podInfo)
 
-	ncgw := net.ParseIP(info.ncGatewayIPAddress)
-	if ncgw == nil {
-		if invoker.ipamMode != util.V4Overlay {
-			return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+		// set result ipconfigArgument from CNS Response Body
+		ip, ncIPNet, err := net.ParseCIDR(info.podIPAddress + "/" + fmt.Sprint(info.ncSubnetPrefix))
+		if ip == nil {
+			return IPAMAddResult{}, errors.Wrap(err, "Unable to parse IP from response: "+info.podIPAddress+" with err %w")
 		}
 
-		ncgw, err = getOverlayGateway(ncipnet)
-		if err != nil {
-			return IPAMAddResult{}, err
+		ncgw := net.ParseIP(info.ncGatewayIPAddress)
+		if ncgw == nil {
+			if (invoker.ipamMode != util.V4Overlay) && (invoker.ipamMode != util.DualStackOverlay) {
+				return IPAMAddResult{}, errors.Wrap(errInvalidArgs, "%w: Gateway address "+info.ncGatewayIPAddress+" from response is invalid")
+			}
+
+			if net.ParseIP(info.podIPAddress).To4() != nil { //nolint:gocritic
+				ncgw, err = getOverlayGateway(ncIPNet)
+				if err != nil {
+					return IPAMAddResult{}, err
+				}
+			} else if net.ParseIP(info.podIPAddress).To16() != nil {
+				ncgw = net.ParseIP(overlayGatewayV6IP)
```

Comment:
yes, we got this fixed ipv6 gateway from HNS team

Reply:
yes, HNS team gave this and we do not have this requirement for now because Jaeryn fixed this before was to fix overlay gateway bug that prevented cloud-node-manager-windows from initializing node because it couldn't reach IMDS. An additional VFP rule was implicitly plumbed by HNS to add 169.254.1.1/16 to the exception list. This is most likely based off the RouteConfiguration in the "azure" HNS network.  This additional VFP rule was skipping SNAT for any IP address in this range 169.254.0.0-169.254.255.255. We'll set the first .1 IP of podcidr as the gateway now.